### PR TITLE
Structurally resolve in `adjust_for_branches`

### PR DESCRIPTION
**Pull request by compiler-errors**
r? lcnr

### CODE REVIEW

**Review by lcnr — COMMENTED** [review]
looking through the api of `Expectation`, there's also `fn resolve` which I believe should also normalize with the new solver and we can probably craft tests for all/mosts its use-sites.

`adjust_for_branches` can then simply match on `self.resolve(fcx, expr.span)` :thinking:

**Comment by compiler-errors** [comment]
@rustbot ready

So, structurally resolving means that we now eagerly process obligations which affects our ability to deduplicate reported predicates. Diagnostics have gotten kinda bad for RPITs in the new solver and with GCE, but I guess we could fix that later.

@bors try @rust-timer queue

File: compiler/rustc_hir_typeck/src/expectation.rs [hunk 1]
```diff
@@ -77,39 +81,31 @@ impl<'a, 'tcx> Expectation<'tcx> {
         }
     }
 
-    /// Resolves `expected` by a single level if it is a variable. If
-    /// there is no expected type or resolution is not possible (e.g.,
-    /// no constraints yet present), just returns `self`.
-    fn resolve(self, fcx: &FnCtxt<'a, 'tcx>) -> Expectation<'tcx> {
+    pub(super) fn structurally_resolve(
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/expectation.rs:84]: I decided to rename these methods to make it clear that they structurally resolve.

**Comment by lcnr** [comment]
r=me with good perf

**Comment by compiler-errors** [comment]
I think perf is insignificant but I guess I could gate the structural resolve behind `-Znext-solver` if @lcnr cares much about it.

**Comment by lcnr** [comment]
nah

@bors r+

**Comment by compiler-errors** [comment]
:(

@bors retry

**Pull request merged**